src/cobra/core/gene.py

```diff
@@ -299,3 +271,360 @@ def _repr_html_(self):
+        """
+        self.update_genes()
+        return frozenset(self._genes)
+
+    def update_genes(self) -> None:
+        """Update genes, used after changes in GPR.
+
+        Walks along the AST tree of the GPR class, and modifies self._genes
+
+        """
+        if hasattr(self, "body"):
+            walker = GPRWalker()
+            walker.visit(self)
+            self._genes = walker.gene_set
+            if "" in self._genes:
+                self._genes.remove("")
+
+    def __eval_gpr(
+        self,
+        expr: Union[Expression, list, BoolOp, Name],
+        knockouts: Union[DictList, set],
+    ) -> bool:
+        """Evaluate compiled ast of gene_reaction_rule with knockouts.
+
+        Parameters
+        ----------
+        expr : Expression or GPR or list or BoolOp or Name
+            The ast of the gene reaction rule
+        knockouts : DictList, set
+            Set of genes that are knocked out
+
+        Returns
+        -------
+        bool
+            True if the gene reaction rule is true with the given knockouts
+            otherwise false
+        """
+        # just always call the recursions as self.__eval_gpr(a, b)
+        if isinstance(expr, Expression) or isinstance(expr, GPR):
+            if not hasattr(expr, "body"):
+                return True
+            return self.__eval_gpr(expr.body, knockouts)
+        elif isinstance(expr, Name):
+            return expr.id not in knockouts
+        elif isinstance(expr, BoolOp):
+            op = expr.op
+            if isinstance(op, Or):
+                # noinspection PyTypeChecker
+                return any(self.__eval_gpr(i, knockouts) for i in expr.values)
+            elif isinstance(op, And):
+                # noinspection PyTypeChecker
+                return all(self.__eval_gpr(i, knockouts) for i in expr.values)
+            else:
+                raise TypeError("unsupported operation " + op.__class__.__name__)
+        elif expr is None:
+            return True
+        else:
+            raise TypeError("unsupported operation  " + repr(expr))
+
+    def eval(self, knockouts: Union[DictList, Set, str, Iterable] = None) -> bool:
+        """Evaluate compiled ast of gene_reaction_rule with knockouts.
+
+        This function calls __eval_gpr, but allows more flexibility in input, including
+        name, and list.
+
+        Parameters
+        ----------
+        knockouts
+            Which gene or genes to knoc out
+
+        Returns
+        -------
+        bool
+            True if the gene reaction rule is true with the given knockouts
+            otherwise false
+
+        """
+        if knockouts is None:
+            knockouts = set()
+        if knockouts is str:
+            knockouts = list(knockouts)
+        if hasattr(self, "body"):
+            return self.__eval_gpr(self.body, knockouts=knockouts)
+        else:
+            return True
+
+    def __ast2str(
```

Comment:
As above, I recommend a single underscore. I will not comment on where calls to this method have to be adapted, too.

```suggestion
    def _ast2str(
```